recipes/lora_finetune_distributed.py

```diff
@@ -307,16 +306,36 @@ def _setup_model(
                 base_model_state_dict_keys=base_model_state_dict.keys(),
             )
 
-            # Load both the base model weights and (if available) the adapter weights. Both
-            # of this should happen only on Rank 0
-            model.load_state_dict(base_model_state_dict, strict=False)
-            if lora_weights_state_dict:
-                model.load_state_dict(lora_weights_state_dict, strict=False)
+        self.adapter_params = get_adapter_params(model)
+        set_trainable_params(model, self.adapter_params)
 
-        else:
-            # For non-zero ranks, load the model on meta device
-            with utils.set_default_dtype(self._dtype), torch.device("meta"):
-                model = config.instantiate(cfg_model)
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                model, auto_wrap_policy={modules.TransformerDecoderLayer}
+            )
+
+        for m in model.modules():
+            if isinstance(m, modules.TransformerDecoderLayer):
+                fully_shard(m)
+        fully_shard(model)
```

Comment:
In FSDP1, we wrap each `TransformerDecoderLayer` and then root `model` as well. It's blackboxed in `auto_wrap_policy=utils.lora_fsdp_wrap_policy(modules_to_wrap={modules.TransformerDecoderLayer})`

In FSDP2, we un-blackboxed it to this for-loop. It you perfer, this can be factored into a util function in torchtune so user call `util.fully_shard(model, modules_to_wrap)` 

Personally I have bias towards un-blackboxed approach since people can modify the for-loop to achieve different wrapping

Reply:
the equivalence SHARD_GRAD_OP in FSDP2 is `reshard_after_forward=False` . Do you want it as a config in .yaml?

`fully_shard(model, reshard_after_forward=False)`

Reply:
thanks for the explanation! I love the un-blackboxed approach here - just needs more documentation and explanation :) After reading the FSDP2 RFC, this became a lot clearer.